Add Scrublet as an external tool

pinin4fjords opened the pull request:
This is a PR attempting to address https://github.com/theislab/scanpy/issues/173, and building on the work of @swolock in his fork (see https://github.com/theislab/scanpy/compare/master...swolock:master). 

I've taken a different approach to that fork:

- Various component functions wrapped individually (core Scrublet functionality, doublet simulation), to allow for custom workflows with different Scanpy preprocessing where required.
- ... but a scrub_doublets() function that does pre-processing analogously with Scrublet (but using Scanpy calls) is what imagine most people will use. 
- I've moved what was sensible to use Scanpy functions.  
- ... but some parts are left alone. In particular the PCA seems like it's used a bit differently in Scrublet, fitting the model to the observed transcriptomes (.fit()) and then applying dimensionality reductions to both observed and simulated matrices (.transform()). That's complex to implement with Scanpy functions (which just uses fit_transform()), so thought I'd leave that in Scrublet (which calls the same libraries anyway). Similarly the neighbour graph generation done in the classifier is quite wired in, so I haven't tried to pull it into Scanpy.

I hope that this is a low-maintenance solution, allowing use of Scrublet with Scanpy functions, without a large amount of re-implementation which might cause drift wrt Scrublet itself. 

It does seem to work, but happy to make any changes you or @swolock want- would just be really useful for us to have this in Scanpy :-).

pinin4fjords committed:
Add Scrublet as an external tool

pinin4fjords committed:
Docstring fixes

pinin4fjords committed:
Apply black formatting

pinin4fjords committed:
Think the CI might want some full stops

giovp commented on scanpy/external/pp/_scrublet.py:
```diff
@@ -0,0 +1,460 @@
+        valid values, see the documentation for annoy (if `use_approx_neighbors`
+        is True) or sklearn.neighbors.NearestNeighbors (if `use_approx_neighbors`
+        is False).
+    normalize_variance : bool, optional (default: True)
```
can you remove the colon and typing everywhere and try again? There should be the need to put them here as they are already present in the arguments maybe?

pinin4fjords replied:
Ahh yes, thank you. I was copying from https://github.com/swolock/scrublet/blob/master/src/scrublet/scrublet.py I think, but will try taking those out.

pinin4fjords replied:
Hmm, no change @giovp

giovp replied:
yes saw it, need to ask seniors then cause I'm also lost

pinin4fjords committed:
strip colons and defaults from parameter docs

giovp commented:
@Koncopd @ivirshup can you have a look at this? CI fails for some weird docs error that we can't nail down.

pinin4fjords committed:
Try imposing line width limit

ivirshup commented on scanpy/external/pp/_scrublet.py:
```diff
@@ -0,0 +1,511 @@
+    transcriptomes and simulated doublets. Works best if the input is a raw
+    (unnormalized) counts matrix from a single sample or a collection of
+    similar samples from the same experiment.
+   
```
```suggestion

```

I think this is what's causing the test failures. I'm not sure that I think the test is looking for the right thing, it's just finding that there's a line that starts with 3 spaces here. There's some other white space in the docs that can get cleaned up too.

ivirshup commented on scanpy/external/pp/_scrublet.py:
```diff
@@ -0,0 +1,511 @@
+    return adata_sim
+
+
+def plot_histogram(
```
If this is going to be exported from a high level, could it get a name that ties it to scrublet? Maybe `sce.pl.scrublet_score_distribution`?

pinin4fjords replied:
Now moved to sce.pl and named as suggested.

ivirshup commented on scanpy/external/pp/_scrublet.py:
```diff
@@ -0,0 +1,511 @@
+    return adata_obs
+
+
+def simulate_doublets(
```
If this function is going to be exported, could its name tie it to scrublet?

pinin4fjords replied:
Now renamed as suggested

ivirshup reviewed (changes requested):
Looks promising!

A couple requests:

### Tests

Could this get some tests? They should cover any logic that's included here, as well as making sure this doesn't throw errors.

### Docs/ linking functions

Since there are multiple `scrublet` based functions, could they link to each other via ["See Also" sections](https://numpydoc.readthedocs.io/en/latest/format.html#sections)?

### Naming

I've addressed this in some comments on the code, but you make sure all the `scrublet` specific functions are easily identifiable as such?

I also don't think it's obvious from the names `scrub_doublets` and `scrublet` what the difference between the functions is.

### Number of functions and use cases

I'm open to disagreement here. I'm just trying to think about if there's a more simple API here.

`scrub_doublets(adata, ...)` is essentially `scrublet(adata, simulate_doublets(adata, ...), ...)` right? Could this be collapsed to one function, where the second argument defaults to `None`?

This may be confusing since these functions would have a large amount of arguments, and you'd want to check and make sure that simulation and normalization arguments weren't passed if the simulated data was provided.

I guess it depends on what you think the workflows are here. I see two primary use cases:

1. I just have counts (simple use case)
  * Doublets should be simulated
  * Identical normalization should be applied
  * Doublets should be called from the original
2. I'm providing my own simulated data (advanced use case)
  * I think we can require the user to have done their own normalization, since they're already going for more customization
  * All that needs to be done is doublet calling

pinin4fjords committed:
Remove problematic spaces

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

pinin4fjords committed:
Move scrublet plot to plotting module, add 'see also', call scrublet calling function from outer function only

pinin4fjords committed:
Merge branch 'add_scrublet_to_external' of github.com:ebi-gene-expression-group/scanpy into add_scrublet_to_external

pinin4fjords commented:
Thanks for all the good feedback @ivirshup - I'll work on it and re-request review when I'm done.

pinin4fjords committed:
Add Scrublet tests

pinin4fjords committed:
Rename _call_doublets

pinin4fjords committed:
Apply black formatting

pinin4fjords commented:
Okay @ivirshup , think I've addressed your comments:

- old 'scrublet' function now not exposed, has become an internal _scrublet_call_doublets (I like it still being separate, makes the logic easier to read). New sce.pp.scrublet now the main exposed function, with scrublet_simulate_doublets() function available for advanced users.
- plot function moved to scanpy/external/pl.py as scrublet_score_distribution().
- functions linked via 'See also' sections.
- tests added for 'scrublet()' and scrublet_simlulate_doublets().

pinin4fjords committed:
Add scrublet install to testing

pinin4fjords commented:
I checked the new tests locally, can't work out why they're not running in the CI here- hopefully it's obvious to you.

ivirshup commented:
> old 'scrublet' function now not exposed, has become an internal _scrublet_call_doublets (I like it still being separate, makes the logic easier to read)

Oh, I think I wasn't clear here. I was thinking that there would be three doublet calling functions:

1. Simulate doublets. Receives count anndata, returns simulated doublet count anndata.
2. Given two anndata objects, one source data, one simulated, call doublets in the source data. It's assumed both objects have already been normalized.
3. The full workflow. Takes an AnnData object with count data, simulates doublets, runs normalization on both, and then calls doublets on the source object. Uses the previous two functions as well as the normalization workflow internally.

The simple use case is just to call function 3. The advanced use case is to use function 2, potentially with data from function 1, or generated some other way. The advanced use case also allows you to use your own normalization. By not giving function 2 the ability to normalize, we cut down on arguments, and have more modular functions.

What do you think of that?

ivirshup commented on scanpy/external/pp/_scrublet.py:
```diff
@@ -0,0 +1,487 @@
+
+def scrublet_simulate_doublets(
+    adata: AnnData,
+    raw_layer: str = 'raw',
```
```suggestion
    layer = None,
```

Then:

```python
X = sc.get._get_obs_rep(adata, layer=layer)
```

This gives you arguments that look like other functions in scanpy, and a default error message.

pinin4fjords commented:
> What do you think of that?

I actually prefer my current layout- it accomplishes your logic without duplication of arguments etc, and from a user perspective I prefer having one main function I can use in the two ways  (one anndata you get 3., Two anndata you get 2). But I'm also very happy for you to tweak things to match your vision!

pinin4fjords committed:
Match parameter with scanpy conventions

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

pinin4fjords committed:
Use _get_obs_rep()

pinin4fjords committed:
Black again...

ivirshup commented on scanpy/tests/external/test_scrublet.py:
```diff
@@ -0,0 +1,51 @@
+
+    # replace assertions by conditions
+    if not set(['predicted_doublet', 'doublet_score']).issubset(adata.obs.columns):
+        errors.append("doublet columns not set in .obs")
```
```suggestion
assert "predicted_doublet" in adata.obs.columns
assert "doublet_score" in adata.obs.columns
```

ivirshup commented on scanpy/tests/external/test_scrublet.py:
```diff
@@ -0,0 +1,51 @@
+        errors.append("doublet columns not set in .obs")
+
+    if len(adata.obs_names[adata.obs.predicted_doublet]) == 0:
+        errors.append("no doublets identified")
```
```suggestion
    assert adata.obs["predicted_doublet"].any(), "Expect some doublets to be identified"
```

This is boolean valued, right?

pinin4fjords replied:
yup

ivirshup commented on scanpy/tests/external/test_scrublet.py:
```diff
@@ -0,0 +1,51 @@
+
+    assert not errors, "errors occured in scrublet testing:\n{}".format(
+        "\n".join(errors)
+    )
```
```suggestion
```

If the above changes are made, this bit can be removed

ivirshup commented on scanpy/tests/external/test_scrublet.py:
```diff
@@ -0,0 +1,51 @@
+    hvg = sc.pp.highly_variable_genes(logged)
+    adata_obs = adata_obs[:, logged.var['highly_variable']]
+
+    adata_sim = sce.pp.scrublet_simulate_doublets(adata_obs, raw_layer='raw')
```
Didn't this `raw_layer` argument get changed?

How is this not causing test failures? I'm a bit worried these tests aren't being run.

pinin4fjords replied:
Argh- sorry. Would probably go through testing anyway, the rogue argument would be ignored and _get_obs_rep() would pick up .X. 

As mentioned above I couldn't work out why the tests aren't running here, was hoping you'd see it. I'll have a dig, but I'm clearly missing something about how this works so I'm not sure I'll be able to fix it by myself.

ivirshup replied:
Oh, sorry, totally missed that message.

ivirshup reviewed (changes requested):
I'll have a think about the API. It might be that my ideal api should just be go into the main library as a more flexible option.

Otherwise, I think there are a few parts of the tests that should change. I'm a little worried they are not being run on CI, since there's one bit that should definitely fail. Could you check that?

pinin4fjords committed:
Update doublet score check

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

pinin4fjords committed:
update predicted doublet check

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

pinin4fjords committed:
Remove section per previous commits

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

pinin4fjords committed:
fix simulate_doublets arg in tests

ivirshup commented on scanpy/tests/external/test_scrublet.py:
```diff
@@ -0,0 +1,47 @@
+
+    # replace assertions by conditions
+assert "predicted_doublet" in adata.obs.columns
+assert "doublet_score" in adata.obs.columns
```
```suggestion
    assert "predicted_doublet" in adata.obs.columns
    assert "doublet_score" in adata.obs.columns
```

Oops, my bad!

pinin4fjords replied:
Ahh, I should have checked

ivirshup commented:
@pinin4fjords the tests aren't running since scrublet isn't actually being installed (I'm surprised the build still worked, apparently this is just a warning?). From the travis logs:

```
203$ pip install .[dev,test,louvain,leiden,magic,scvi,harmony,skmisc,scrublet]
204Processing /home/travis/build/theislab/scanpy
205  Installing build dependencies ... done
206  Getting requirements to build wheel ... done
207    Preparing wheel metadata ... done
208  WARNING: scanpy 0.1.dev67+g3918588 does not provide the extra 'scrublet'
```

You'll need to add a scrublet entry to `extras_require` here: https://github.com/theislab/scanpy/blob/d56d6beacdd951a010bb6a93078db26e1ac904b0/setup.py#L31-L56

ivirshup commented:
Also sorry about these horrible CI wait times! Hopefully these will be better sometime next week.

pinin4fjords committed:
Fix indent

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

pinin4fjords committed:
Add scrublet to extras_require

pinin4fjords commented:
> @pinin4fjords the tests aren't running since scrublet isn't actually being installed (I'm surprised the build still worked, apparently this is just a warning?). From the travis logs:
> You'll need to add a scrublet entry to `extras_require` here:

Ahh, I see- thank you!

pinin4fjords commented:
> Also sorry about these horrible CI wait times! Hopefully these will be better sometime next week.

Maybe you're having issues with the new Travis credits system? We've certainly had issues- we're gradually migrating to GitHub actions now.

ivirshup commented:
> Maybe you're having issues with the new Travis credits system?

That's my assumption. This project is also somehow still on travis-ci.org. We're working on moving to azure.

pinin4fjords committed:
Import _get_obs_rep correctly

pinin4fjords committed:
Black again...

pinin4fjords commented:
Tests passing! With exception of silly Black formatting issue I've just corrected.

pinin4fjords committed:
yet more black

ivirshup commented on scanpy/external/pp/_scrublet.py:
```diff
@@ -0,0 +1,486 @@
+        ],
+        'n_neighbors': n_neighbors,
+        'random_state': random_state,
+    }
```
```suggestion
    adata_obs.uns['scrublet'] = {
        'threshold': scrub.threshold_
        'doublet_scores_sim': scrub.doublet_scores_sim_
        'doublet_parents': adata_sim.obsm['doublet_parents']
        'parameters': {
            'expected_doublet_rate': expected_doublet_rate,
            'sim_doublet_ratio': (
                adata_sim.uns
                .get('scrublet', {})
                .get('parameters', {})
                .get('sim_doublet_ratio', None)
            ),
            'n_neighbors': n_neighbors,
            'random_state': random_state,
        }
    }
```

Simplifies the assignment a bit, and allows `adata_sim` to not have `scrublet` parameters

ivirshup commented on scanpy/external/pp/_scrublet.py:
```diff
@@ -0,0 +1,486 @@
+    adata_sim.obsm['doublet_parents'] = scrub.doublet_parents_
+    adata_sim.uns['scrublet'] = {}
+    adata_sim.uns['scrublet']['parameters'] = {}
+    adata_sim.uns['scrublet']['parameters']['sim_doublet_ratio'] = sim_doublet_ratio
```
```suggestion
    adata_sim.uns['scrublet'] = {'parameters': {'sim_doublet_ratio': sim_doublet_ratio}}
```

ivirshup commented on scanpy/external/__init__.py:
```diff
@@ -34,6 +34,11 @@
+Multiplet detection
+~~~~~~~~~~~~~~~~~~~
+
+   pp.scrub_doublet
```
```suggestion
   pp.scrublet
   pp.scrublet_simulate_doublets
   pl.scrublet_score_distribution
```

Looks like this didn't get updated

ivirshup reviewed (commented):
Sorry for the long turn around!

A couple minor changes, and a release note is needed, then should be good to go.

ivirshup commented:
@pinin4fjords, just a couple minor changes requested. I'd add them myself, but it looks like I don't have permissions to push to this branch on your repo.

pinin4fjords committed:
Simplifies the assignment a bit, and allows adata_sim to not have scrublet parameters

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

pinin4fjords committed:
adata_sim parameter simplfication

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

pinin4fjords committed:
Apply forgotten updates

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

pinin4fjords committed:
Merge branch 'master' into add_scrublet_to_external

pinin4fjords commented:
Thanks @ivirshup ! I'm on xmas leave, but I've merged your changes without looking too carefully! Have also added you as collaborator on our fork so you can make further changes if required.

ivirshup commented:
Enjoy your Holidays!

No worries! I'm still having some trouble pushing to this, so I'm just going to merge it through a separate PR that points to this one. Thanks for working on this!
